leif/github_repository.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leif
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	goGH "github.com/google/go-github/v32/github"
+)
+
+const sloConfigFileName = "issue_slo_rules.json"
+
+type notAFileError struct {
+	path string
+	org  string
+	repo string
+}
+
+func (e *notAFileError) Error() string {
+	return fmt.Sprintf("The path %v in %v/%v does not correspond to a file", e.path, e.org, e.repo)
+}
+
+type noContentError string
+
+func (e *noContentError) Error() string {
+	return string(*e)
+}
+
+// githubRepoService is an interface defining the needed behaviour of the GitHub client
+// This way, the default client may be replaced for testing
+type gitHubRepoService interface {
+	GetContents(ctx context.Context, owner, repo, path string, opts *goGH.RepositoryContentGetOptions) (fileContent *goGH.RepositoryContent, directoryContent []*goGH.RepositoryContent, resp *goGH.Response, err error)
+}
+
+// githubClient is a a wrapper around the GitHub client's RepositoriesService
+type gitHubClient struct {
+	Repositories gitHubRepoService
+}
+
+// NewGithubClient creates a wrapper around the GitHub client's RepositoriesService
+// The RepositoriesService can be replaced for unit testing
+func NewGitHubClient(httpClient *http.Client, repoMock gitHubRepoService) gitHubClient {
+	if repoMock != nil {
+		return gitHubClient{
+			Repositories: repoMock,
+		}
+	}
+	client := goGH.NewClient(httpClient)
+
+	return gitHubClient{
+		Repositories: client.Repositories,
+	}
+}
```

Comment:
I think it would be best to include a comment that links to the explanation of where this inspiration came from.

It also is probably best to split the GitHub client (and associate methods) into either 

1. A separate file
2. A separate package

for better isolation and testing.